models/image.py

```diff
@@ -877,72 +702,110 @@ def from_images(cls, images):
         # Note that "data" is not filled by this method, also the provenance is empty!
         return output
 
     @classmethod
-    def from_ref_and_new(cls, ref, new):
+    def from_ref_and_new(cls, ref_image, new_image):
+        return cls.from_new_and_ref(new_image, ref_image)
+
+    @classmethod
+    def from_new_and_ref(cls, new_image, ref_image):
         """
         Create a new Image object from a reference Image object and a new Image object.
         This is the first step in making a difference image.
+
         The output image doesn't have any data, and is created with
         nofile=True. It is up to the calling application to fill in the
         data, flags, weight, etc. using the appropriate preprocessing tools.
+
+        The Image objects used as inputs must have their own data products
+        loaded before calling this method, so their provenances will be recorded.
+        The provenance of the output object should be generated, then a call to
+        output.provenance.upstreams = output.get_upstream_provenances()
+        will make sure the provenance has the correct upstreams.
+
         After that, the data needs to be saved to file, and only then
         can the new Image be added to the database.
 
         Parameters
         ----------
-        ref: Image object
-            The reference image to use.
-        new: Image object
+        new_image: Image object
             The new image to use.
+        ref_image: Image object
+            The reference image to use.
 
         Returns
         -------
         output: Image
             The new Image object. It would not have any data variables or filepath.
         """
+        if ref_image is None:
+            raise ValueError("Must provide a reference image.")
+        if new_image is None:
+            raise ValueError("Must provide a new image.")
+
         output = Image(nofile=True)
 
         # for each attribute, check the two images have the same value
-        for att in ['section_id', 'instrument', 'telescope', 'filter', 'project', 'target']:
-            ref_value = getattr(ref, att)
-            new_value = getattr(new, att)
+        for att in ['instrument', 'telescope', 'project', 'section_id', 'filter', 'target']:
+            ref_value = getattr(ref_image, att)
+            new_value = getattr(new_image, att)
 
             if att == 'section_id':
                 ref_value = str(ref_value)
                 new_value = str(new_value)
-            if ref_value != new_value:
+
+            # TODO: should replace this condition with a check that RA and Dec are overlapping?
+            #  in that case: what do we consider close enough? how much overlap is reasonable?
+            #  another issue: what happens if the section_id is different, what would be the
+            #  value for the subtracted image? can it live without a value entirely?
+            #  the same goes for target. what about coadded images? can they have no section_id??
```

Comment:
coadded images may well be coadded from different section ids, so the section id for a coadded image is not well defined.  We need to make sure that the image infrastructure is able to handle that, which may require some work.  So, we should only check that the filter is the same.

Re: RA and Dec overlapping, I'd say if they have at least (say) 5%, or even 1%, overlap, go ahead and do the subtraction.  Somebody might be trying to find something specific they know is in that overlap region.  The actual subtraction algorithm itself may fail if there's not enough overlap to psf match, for instance, but at this high-level definition of the subtraction, we shouldn't worry about that.  That will be the problem of the implementation of the subtraction algorithm.

Reply:
Issue #179